src/apps/service/apps/de/listings.clj

```diff
@@ -297,27 +298,32 @@
     (svc-util/format-job-stats app admin?)
     app))
 
-(defn- format-app-listing
-  "Formats certain app fields into types more suitable for the client."
-  [admin? perms beta-ids-set public-app-ids {:keys [id] :as app}]
-  (-> (assoc app :can_run (app-can-run? app))
-      (dissoc :tool_count :task_count :external_app_count :lower_case_name)
-      (format-app-listing-job-stats admin?)
-      (format-app-ratings)
-      (format-app-pipeline-eligibility)
-      (format-app-permissions perms)
-      (assoc :can_favor true :can_rate true :app_type "DE" :system_id c/system-id)
-      (assoc :beta (contains? beta-ids-set id))
-      (assoc :is_public (contains? public-app-ids id))
-      (remove-nil-vals)))
-
 (defn- app-ids->beta-ids-set
   "Filters the given list of app-ids into a set containing the ids of apps marked as `beta`"
   [username app-ids]
   (let [beta-avu {:attr (workspace-metadata-beta-attr-iri)
                   :value (workspace-metadata-beta-value)}]
     (set (metadata-client/filter-by-avus username app-ids [beta-avu]))))
 
+(defn- get-app-listing-formatter
+  "Returns a function that can be used to format the listing for a single app. Using a higher order function for this
+   makes it easier to consolidate repetitive tasks without degrating performance."
```

Comment:
Could you explain this more? I don't have any problem with the change, but I wasn't sure why this approach was better than using `partial` like before, so I'm wondering what I'm not understanding.

Reply:
If I continued using `partial` like we were before, I would have had to add yet another parameter to the formatting function, which already has a lot of positional parameters. I thought briefly about changing the positional parameters to named parameters (and putting all of the named parameters into an option map so that the app listing itself is still a positional parameter). Ultimately, however, I decided to go with this approach because I also would have had to add code to obtain the limit checks to every place where the formatting function is being called. Obtaining the limit checks and the list of beta app IDs here, allows me to remove the (minor) amount of duplicate code from all of the app listing functions. There are trade-offs, of course. The returned anonymous function is another level of indirection to think about. I still think this helps to make the code a little bit easier to read, though.

Reply:
Makes sense. I think my mind got caught up on the mention of performance -- this and using `partial` are probably similar performance-wise, but this is definitely easier to read, and more performant than other ways of making it more easily readable, so I think I've connected the dots. Thanks!

Reply:
Ah, yeah, I guess I didn't really need to mention performance. All I really meant was that we'd still be retrieving the beta ID list and the limit checks once even though those things aren't being done in the calling functions explicitly.